File: ga/gpt3_5_result_2024_6/result/1968G2_gpt/15_1968G2.c
use std::io::{self, BufRead};

const N: usize = 200000;
const N_: usize = 1 << 18;

fn max(a: usize, b: usize) -> usize {
    if a > b {
        a
    } else {
        b
    }
}

fn build(aa: &mut [usize], n: usize) -> Vec<usize> {
    let mut st = vec![0; N_ * 2];
    let mut n_ = 1;
    
    while n_ < n {
        n_ <<= 1;
    }
    
    st[n_..n_ + n].copy_from_slice(aa);
    
    for i in (1..n_).rev() {
        st[i] = max(st[i << 1], st[i << 1 | 1]);
    }
    
    st
}

fn query(st: &[usize], n_: usize, l: usize, a: usize) -> usize {
    let mut l = l + n_;
    let mut r = n_ - 1;
    
    while l <= r {
        if (l & 1) == 1 {
            if st[l] >= a {
                let mut temp = l;
                while temp < n_ {
                    temp = if st[temp << 1] >= a {
                        temp << 1
                    } else {
                        temp << 1 | 1
                    };
                }
                return temp - n_;
            }
            l += 1;
        }
        l >>= 1;
        r >>= 1;
    }
    
    n_
}

fn main() {
    let stdin = io::stdin();
    
    let mut input = String::new();
    stdin.lock().read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();
    
    for _ in 0..t {
        input.clear();
        stdin.lock().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace();
        let n: usize = iter.next().unwrap().parse().unwrap();
        let kl: usize = iter.next().unwrap().parse().unwrap();
        let kr: usize = iter.next().unwrap().parse().unwrap();
        let cc = iter.next().unwrap();
        
        let mut zz = vec![0; n];
        let cc_chars: Vec<char> = cc.chars().collect();
        
        let mut l = 0;
        let mut r = 0;
        
        for i in 1..n {
            if i + zz[i - l] < r {
                zz[i] = zz[i - l];
            } else {
                l = i;
                r = max(r, i);
                while r < n && cc_chars[r] == cc_chars[r - l] {
                    r += 1;
                }
                zz[i] = r - l;
            }
        }
        
        let st = build(&mut zz, n);
        let mut ans = vec![0; n + 1];
        
        for p in 1..=n {
            let mut k = 0;
            let mut i = 0;
            
            while i < n {
                i = query(&st, n_, i + p, p);
                if i < n_ {
                    k += 1;
                }
            }
            ans[k] = max(ans[k], p);
        }
        
        for k in (1..n).rev() {
            ans[k] = max(ans[k], ans[k + 1]);
        }
        
        for k in kl..=kr {
            print!("{} ", ans[k]);
        }
        println!();
    }
}

File: ga/glm4_result_2024_6/result/1968G2_cpp_glm/15_1968G2.rs
use std::io;
use std::io::{stdout, Write};

const N: usize = 2e5 as usize + 3;
const T: usize = 1;

fn main() {
    let mut stdout = stdout();
    let n: usize;
    let x: usize;
    let y: usize;
    let mut a = String::new();
    let mut z = vec![0; N];
    let mut p = vec![0; N];
    let mut t = vec![0; N];
    let mut ans = vec![0; N];

    for _ in 0..T {
        io::stdin().read_line(&mut a).unwrap();
        let a: Vec<char> = a.trim().chars().collect();
        let n = a.len();
        for i in 1..=n {
            p[i] = i;
            t[i] = i;
            ans[i] = 0;
            z[i] = 0;
        }
        z[1] = n;

        for i in 2..=n {
            let mut j = 0;
            while j + z[j] > i {
                z[i] = std::cmp::min(z[i - j + 1], j + z[j] - i);
            }
            while a[i + z[i]] == a[1 + z[i]] {
                z[i] += 1;
            }
            if i + z[i] > j + z[j] {
                j = i;
            }
        }

        t.sort_by(|i, j| z[*i] < z[*j]);

        for i in 1..=n {
            let mut k = 1;
            while k <= n && z[t[k]] < i {
                p[t[k]] = t[k] + 1;
                k += 1;
            }
            let mut c = 0;
            for j in 1..=n {
                let mut j = Find(j + i);
                c += 1;
            }
            ans[c] = i;
        }

        for i in (0..n).rev() {
            ans[i] = std::cmp::max(ans[i], ans[i + 1]);
        }

        for i in x..=y {
            write!(stdout, "{} ", ans[i]).unwrap();
        }
        write!(stdout, "\n").unwrap();
    }
}

fn Find(k: usize) -> usize {
    if k > N || p[k] == k {
        return k;
    }
    p[k] = Find(p[k]);
    p[k]
}